### PR TITLE
Fix field labels and deployment flow model name UX.

### LINKED ISSUE #2081: The behaviour of the float label
If the field is empty, the label should be 16px and then shrink to 12px when clicking into the field and adding information. 

float lable exploration

### LINKED ISSUE #2096: DF: Default model/cred names not cleared on focus.
Automatically clear the default name and save our users a bit o'typing.

### PR DESCRIPTION
**Pull request by kadams54**
Fixes #2096.
Fixes #2081.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5951/
Test FAILed.

**Comment by kadams54** [comment]
## QA

Works best if you use the guiproxy tool to attach to a real backend. Either that or do the `make dist && juju upgrade-gui` dance.

1. Go to your user profile.
2. Click on the "Create a model" button.
3. Add an application to the canvas.
4. Deploy.

In the deployment flow, the first screen will have your model's name selected so that you can just start typing a new name if you so desire.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5960/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5961/
Test PASSed.

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 1]
```diff
@@ -529,6 +529,17 @@ YUI.add('deployment-flow', function() {
     },
 
     /**
+      Select the value of the model name input, for easier replacement with a
+      different name.
+
+      @method _generateModelNameSection
+      @returns {Object} The markup.
+    */
+    _selectModelName: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: This is no longer used

File: jujugui/static/gui/src/app/components/generic-input/generic-input.js [hunk 2]
```diff
@@ -38,7 +39,7 @@ YUI.add('generic-input', function() {
     },
 
     getInitialState: function() {
-      return {errors: null};
+      return {errors: null, focus: false};
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/generic-input/generic-input.js]: s/focus/focused as it's a state not an action

Comment (kadams54) [inline jujugui/static/gui/src/app/components/generic-input/generic-input.js]: "focus" was to remain consistent with the CSS class that the state toggles on and off. That CSS class is "focus" to be consistent with the `:focus` pseudo selector. In this case, when 'input:focus' is in effect, we want the label element to also have a `focus` class, which we achieve by having the input's onFocus event handler set the `focus` attribute in state.

All that said, I went back and forth between "focus" and "focused" and don't care that strongly. I gave a slight edge to "focus" for the reason above, but would be fine with changing it.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5965/
Test FAILed.

**Comment by kadams54** [comment]
Build failure seems to be spurious; a manual build (http://ci.jujugui.org:8080/job/juju-gui/5968/) fixed it.

**Comment by anthonydillon** [comment]
Code look ok thanks :+1:

**Comment by anthonydillon** [comment]
Thanks @kadams54 looking good but seems some of the styling is a little off. Could you match the spec here: https://github.com/juju/juju-gui/issues/2081#issuecomment-252289835.

**Comment by kadams54** [comment]
@anthonydillon Made the color/size changes. That said, I ended up using `$warm-grey` for the starting color, which is `#888888`, instead of `#666666`, as that is not a pre-set color in vanilla land.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5970/
Test FAILed.

**Comment by anthonydillon** [comment]
Thanks, looks good. Small suggestion about the transition but its your call @kadams54

File: jujugui/static/gui/src/app/components/generic-input/generic-input.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/generic-input/generic-input.js]: Keeping "focus" is good enough I think.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5976/
Test PASSed.

File: jujugui/static/gui/src/app/components/generic-input/_generic-input.scss [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
-        // Needs to be relative to appear above the input.
-        position: relative;
-        margin: 4px 0 -23px 10px;
-        padding-top: 5px;
-        font-size: 0.688em;
+        // Needs to be positioned to appear above the input.
+        color: $cool-grey;
+        left: 10px;
+        position: absolute;
+        top: 25px;
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/generic-input/_generic-input.scss]: To center the label to more closely match the design when the input is not selected this should probably be ~17px.

Comment (hatched) [inline jujugui/static/gui/src/app/components/generic-input/_generic-input.scss]: This label also needs a `user-select: none` as I've selected the label multiple times. Or maybe the label should be behind the input? I'll leave that fix up to you.

Comment (kadams54) [inline jujugui/static/gui/src/app/components/generic-input/_generic-input.scss]: Switched to 17px and `user-select: none`. With both in place it looks much closer to the animation.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5982/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5983/
Test PASSed.

**Comment by anthonydillon** [comment]
Sorry I missed it before but looks like the colors are inverted. It should be grey when large and get darker when focused.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5985/
Test PASSed.

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**